### PR TITLE
Add orderColumn to repeatable to work with pivot orders

### PR DESCRIPTION
**Pull request by pxpm**
### BEFORE - What was wrong? What was happening before this PR?

Repeatable relations were not orderable because we needed a way for repeatable to have a "order" field, and that order field should also be respected when displaying the results.

### AFTER - What is happening after this PR?

Developer can provide `'orderColumn' => 'order_column_name'` and backpack will create an hidden input for that column and keep it updated with current order in table. 

### How did you achieve that, in technical terms?

Hidden input for each row adding the "order column" and keeping it in sync with table orders. When getting the field for display we make sure that the order still apply. 

It's debatable here if Backpack should return the ordered results, or if developer should `order by column` in their relationship definition. 

I opted by beeing Backpack handling that as it would not force to have the relation ALLWAYS ordered by some column, so we can have the relationship non-ordered by default (in the relationship definition), and still have the ability to work with some database order that could be needed for some specific cases but not by general relationship. 

Rephrasing, I think if developer WANTS to add `->order` for their relationship they can still do it, but if they don't do it, backpack will still work.

### Is it a breaking change or non-breaking change?

It's a non breaking!

### How can we test the before & after?

Before orders were completely disabled for pivot relations so .. no test.
After you need to add the `order_column` to your database, and setup that column in: `orderColumn` for that pivot relation. 
Don't forget that you need to add the `order_column` to the `->withPivot()` statement in your relationship definition.

```php
CRUD::field('dummyproducts')->subfields([
            [
                'name' => 'notes'
            ], 
        ])->orderColumn('order'); //order is the name of the column in the pivot table
```

Let me know if you think we are on the rigth track here @tabacitu

### CODE REVIEW

**Comment by tabacitu** [comment]
> It's debatable here if Backpack should return the ordered results, or if developer should order by column in their relationship definition. I opted by beeing Backpack handling that as it would not force to have the relation ALLWAYS ordered by some column, so we can have the relationship non-ordered by default (in the relationship definition), and still have the ability to work with some database order that could be needed for some specific cases but not by general relationship.

Totally agree. It would be easier for the dev to NOT have to add a certain order to their relationship. The field will do it for them.

> Before orders were completely disabled for pivot relations so .. no test. After you need to add the order_column to your database, and setup that column in: orderColumn for that pivot relation. Don't forget that you need to add the order_column to the ->withPivot() statement in your relationship definition.

This should also work for `hasMany` relationships, with no pivot table, correct (eg. `Invoice hasMany InvoiceItem`)? And in that case, there is no need to mention the `order` on the relationship.

> Let me know if you think we are on the rigth track here @tabacitu

Diving into the code right now.

File: src/resources/views/crud/fields/inc/repeatable_row.blade.php [hunk 1]
```diff
@@ -16,6 +21,7 @@
         </button>
         @endif
     </div>
+    @if($field['reorder'] && isset($field['orderColumn']) && $field['orderColumn'] !== false)<input type="hidden" class="order_hidden_input" name="{{$reorder_input_name}}" value="" />@endif
```
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/resources/views/crud/fields/inc/repeatable_row.blade.php:24]: Here's what I propose, to solve both issues in an understandable way. Let's start from the docs. **This is a `repeatable` feature, called `reorder`**. It extends the previous `reorder` functionality in `repeatable`, to make it possible to reorder not just _visually_, but also update a hidden field, when that up/down arrows are pressed. Correct?

How do you add a hidden `reorder` field to a `repeatable`? Let's look at the "modified docs" for the `repeatable` field (look at the bottom):

```php
[   // repeatable
    'name'  => 'testimonials',
    'label' => 'Testimonials',
    'type'  => 'repeatable',
    'subfields' => [
        [
            'name'    => 'name',
            'type'    => 'text',
            'label'   => 'Name',
            'wrapper' => ['class' => 'form-group col-md-4'],
        ],
        [
            'name'    => 'position',
            'type'    => 'text',
            'label'   => 'Position',
            'wrapper' => ['class' => 'form-group col-md-4'],
        ],
        [
            'name'    => 'company',
            'type'    => 'text',
            'label'   => 'Company',
            'wrapper' => ['class' => 'form-group col-md-4'],
        ],
        [
            'name'  => 'quote',
            'type'  => 'ckeditor',
            'label' => 'Quote',
        ],
    ],

    // optional
    'new_item_label'  => 'Add Group', // customize the text of the button
    'init_rows' => 2, // number of empty rows to be initialized, by default 1
    'min_rows' => 2, // minimum rows allowed, when reached the "delete" buttons will be hidden
    'max_rows' => 2, // maximum rows allowed, when reached the "new item" button will be hidden
    'reorder' => false, // hide up&down arrows next to each row (no reordering)
    'reorder' => true, // show up&down arrows next to each row
    'reorder' => 'order', // show arrows and adds a hidden subfield with that name (value gets updated when rows move)
    'reorder' => ['name' => 'order', 'type' => 'number', 'attributes' => ['data-reorder-input' => true]], // shows arrows and add a visible number subfield, which will be updated when rows move up and down
],
```

When starting from the docs we see that:
- rather than calling it "order" it makes more sense to call it "reorder" because we already have that notion;
- rather than defining a "column" it makes more sense to define a "subfield"; but that would be odd, and we'd have to use underscore to match the other features inside `repeatable`... so `reorder_subfield` which is a bit ugly so... why not use `reorder` itself? in the end, if `reorder => false` the hidden field doesn't make sense anyway;

I think this makes the syntax more consistent with the other features inside `repeatable`, and easy to understand. You can define a few stuff to `reorder`:
- true/false
- a string => hidden input
- an array => that input

This means a dev can do:

```php
CRUD::field('testimonials')->subfields([
    'title',
    ['name' => 'body', 'type' => 'textarea'],
])->reorder('order'); // will add a hidden subfield for it

// or if you want something different about that subfield:

CRUD::field('testimonials')->subfields([
    'title',
    ['name' => 'body', 'type' => 'textarea'],
])->reorder([
    'name' => 'order', 
    'type' => 'text', 
    'attributes' => ['data-reorder-input' => true]
]); // if you want to customize something about it (notice I've made it "text" here for some reason)
```

Or manually (which nobody will do):

```php
CRUD::field('testimonials')->subfields([
    'title',
    ['name' => 'body', 'type' => 'textarea'],
    ['name' => 'order', 'type' => 'hidden', 'attributes' => ['data-reorder-input' => true]],
]);
```

All of the above will work. But most people will just do either `reorder(false)` or `reorder('order')`.

----

How this works is using JS. The `repeatable` field sees inputs with `data-reorder-input`, so it knows (inside its JS) to:
- reorder the items by that field, on field initialization;
- when items are reordered, to update all `input[type=hidden][data-reorder-input=true]` inputs in that `repeatable` with the new order values;

----

So sum up, what I'm proposing:
- [ ] let's use `->reorder()` instead of `->orderColumn()`
- [ ] let's add a real subfield, if given by `reorder`:
     - [ ] string => a hidden field with `data-reorder-input=true`
     - [ ] array => that given field
- [ ] let's move the reordering of the rows to JS (no point in Backpack having any notion of the `$subfield['reorder']`, it can be purely a feature offered by that field type;

--- 

What do you think @pxpm ?

**Review by tabacitu — CHANGES_REQUESTED** [review]

File: src/resources/views/crud/fields/inc/repeatable_row.blade.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (pxpm) [reply src/resources/views/crud/fields/inc/repeatable_row.blade.php:24]: @tabacitu it's the same problem as with `subfields/pivotFields`, we cannot support it so down in the chain only, because otherwise we have no "reorder_column" in the `pivot fields` when we want to get the values. 

Bare with me:

You setup in controller the field for the `BelongsToMany` with pivot where you want to have a reorder field:

```php
CRUD::field('postalboxes')->subfields([
            [
                'name' => 'postal_name'
            ], 
        ])->reorder('order_column');
```

From here we will reach the `relationship.blade.php`, and from there we would endup with `entries.blade.php`. Only in `entries` we will setup the order field and add it to subfields:

```php
// from entries.blade.php
switch ($field['relation_type']) {
        case 'MorphToMany':
        case 'BelongsToMany':
            $field['subfields'] = Arr::prepend($field['subfields'], $pivotSelectorField);
            break;
        case 'MorphMany':
        case 'HasMany':
            if(isset($entry)) {
                $field['subfields'] = Arr::prepend($field['subfields'], [
                    'name' => $entry->{$field['name']}()->getLocalKeyName(),
                    'type' => 'hidden',
                ]);
            }
            break;
        default: {
            if(is_string($field['reorder'])) {
                $field['subfields'] = Arr::prepend($field['subfields'], [
                    'name' => $field['reorder'],
                    'type' => 'hidden',
                ]);
            }
        }
        break;
    }
```

When you go to the edit page, running the "crud" functions to get the values from database you do:
- get relationship fields from crud fields (your `postalboxes` field will be returned in this example)
- now get the `subfields` from that relationship and return the values (at this time, **the only subfield is `postal_name`**, because we only setup the ´order_column` field after in the blades files. 

So to sum up:
- I agree with you using only `reorder` key for bool/column definition 👍 
- I agree with you that we could add it as a subfield instead of manually define the hidden input 👍 
- I don't think we can make it work if upstream crud functions are not aware at all of this functionality. 👎 

I've refactored to accomodate some of your concerns but could not remove the crud awareneess for reorder when dealing with certain relationships.

**Comment by pxpm** [comment]
@tabacitu I've addressed some of your concerns. please have a look at https://github.com/Laravel-Backpack/CRUD/pull/4084#discussion_r784952600 and let me know if you have other suggestions.

File: src/resources/views/crud/fields/relationship/entries.blade.php [hunk 2]
```diff
@@ -11,6 +11,7 @@
     //each row represent a related entry in a database table. We should not "auto-add" one relationship if it's not the user intention.
     $field['init_rows'] = 0;
     $field['subfields'] = $field['subfields'] ?? [];
+    $field['orderColumn'] = $field['orderColumn'] ?? false;
```
Comment (pxpm) [inline src/resources/views/crud/fields/relationship/entries.blade.php:14]:
```suggestion
```

Comment (pxpm) [inline src/resources/views/crud/fields/relationship/entries.blade.php]:
```suggestion
```

File: src/resources/views/crud/fields/relationship/entries.blade.php [hunk 3]
```diff
@@ -25,6 +25,23 @@
...
+                    'data-reorder-input' => true
+                ]
+            ]);
+        }
+        break;
+        case 'array' : {
+            $field['subfields'] = Arr::prepend($field['subfields'], $field['reorder']);
+        }
+        break;
+    }
```
Comment (tabacitu) [inline src/resources/views/crud/fields/relationship/entries.blade.php:44]: Why not move this to `repeatable`? It's his feature, after all. 

The way I see it, if we do move this to `repeatable`, we open up the possibility for somebody to use this feature in `repeatable` with a custom saving process, without a relationship.

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 4]
```diff
@@ -327,7 +327,7 @@ private function createManyEntries($entry, $relation, $relationMethod, $relation
 
         foreach ($items as $item) {
             if (isset($item[$relation_local_key]) && ! empty($item[$relation_local_key])) {
-                $entry->{$relationMethod}()->updateOrCreate([$relation_local_key => $item[$relation_local_key]], Arr::except($item, $relation_local_key));
+                $entry->{$relationMethod}()->where($relation_local_key, $item[$relation_local_key])->update(Arr::except($item, $relation_local_key));
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Create.php:330]: Why is this change needed? I'm a bit weary because I see `updateOrCreate()` has turned into `update()`.

File: src/resources/views/crud/fields/repeatable.blade.php [hunk 5]
```diff
@@ -8,8 +8,22 @@
...
+         case 'string': {
+            usort($field['value'], fn($a, $b) => $a[$field['reorder']] <=> $b[$field['reorder']]);
+         }
+         break;
+         case 'array': {
+            usort($field['value'], fn($a, $b) => $a[$field['reorder']['name']] <=> $b[$field['reorder']['name']]);
+         }
+         break;
+     }
+  }
```
Comment (tabacitu) [inline src/resources/views/crud/fields/repeatable.blade.php:25]: Nice solution reordering in PHP on initial load! Much cleaner than in JS. But this bit threw a huge error for me:

Screenshot 2022-01-15 at 15 40 36

I have `order` and `role` columns on `owner_pet`. In PetCrudController I've added:
```php
        CRUD::field('owners')->subfields([
            ['name' => 'role', 'type' => 'text']
        ])->reorder('order');
```

File: src/resources/views/crud/fields/inc/repeatable_row.blade.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/resources/views/crud/fields/inc/repeatable_row.blade.php]: Ah, ok, I get it now, thank you @pxpm . Maybe the problem isn't that we don't have `order` in `subfields` by then (because it's added in the view). Maybe the problem is that for subfields... we take each attribute individually. I don't see why we would do that, when we can just take ALL attributes in one go. I'm opening a PR to explain this better.

[the remark this answers is not in the document]
Comment (tabacitu) [reply src/resources/views/crud/fields/inc/repeatable_row.blade.php]: Done. Please see https://github.com/Laravel-Backpack/CRUD/pull/4086/files and if that's feasible, let's do that. My tests show it works the same, but I might be missing something.

**Comment by tabacitu** [comment]
Answered! Thank you for your patience @pxpm . Please see my answer in https://github.com/Laravel-Backpack/CRUD/pull/4084#discussion_r785309769 too please (that thread doesn't show up in the diff).

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 4 above)
Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Create.php:330]: This is the correct, it would never be created here because if `!empty($item[$relation_local_key])` means we have an entry to update, so ... it gets created in the `else` condition.

**Comment by tabacitu** [comment]
Tested again, works like a charm. Added docs for it in https://github.com/Laravel-Backpack/docs/commit/9e15a6aadc80056c25bc8b6efa2070d003298e45
